### DISCUSSION TITLE
Version 2.1 broken cluster?

### DISCUSSION THREAD

**Discussion by rkozlo** [apache/kvrocks#910] [Q&A] [answered]
Hi there,
testing latest stable version 2.1. Trying to set cluster but i'm stuck. Everything seems to be correct but for some reasons i have an error
`CLUSTERX SETNODES  "ae23b94ccaf714337e4ce5ba99ef3dc257e300df 10.32.68.251 6666 master - 0-5460 32d332da761f44df7959e5887b6b94cb4667d781 10.32.68.250 6666 master - 5461-10992 862a51f8b6294a4b0729a7c5c929bfd67068d962 10.32.68.249 6666 master - 10993-16383" 1`
and error
`(error) Slot distribution is overlapped`
Changing slots doesn't change a thing.
Hit this command with only one node passes without problem. Is it some bug or unfinished code? DId everything basing on [doc](https://kvrocks.apache.org/docs/Cluster/kvrocks-cluster-introduction). Didn't find any related issues.

**Comment by git-hulk** [comment, marked as answer]
You can use the `redis-cli cluster slots` to have a look at current slots in node. Current implementation have a bug that it can't set the slot to node even the current owner is itself.

**Reply by rkozlo** [reply]
Thanks for reply.
You mean cluster --create? It says node is not running in cluster mode
`[ERR] Node 10.32.68.251:6666 is not configured as a cluster node.`
```
127.0.0.1:6666> config get cluster-enabled
1) "cluster-enabled"
2) "yes"
```
`redis-cli --cluster create 10.32.68.251:6666 10.32.68.250:6666 10.32.68.249:6666`
My config mostly uses default values. Compared with config from try_cluster in repo and don't see a point where my config may be broken.

**Reply by git-hulk** [reply]
No, what I mean is to use `cluster slots`(through redis-cli) to have a look at the current slots info in the Kvrocks node. BTW, Kvrocks should listen on IP like `10.32.68.251` instead of the loopback device when you are in cluster mode.

**Reply by rkozlo** [reply]
Ah ok, but i can't use cluster slots until cluster will not be initialized
`
(error) CLUSTERDOWN The cluster is not initialized
`
for now it is only sandbox, trying to do first run and it is listening on 0.0.0.0 just for tests

**Reply by git-hulk** [reply]
Ooooops, need to use the specified IP like 10.32.68.251 instead of 0.0.0.0, should fix this issue later.

**Reply by rkozlo** [reply]
It is host ip and kvrocks is running inside container. So i set static ip for container and bind on this ip
```
10.88.0.117:6666> config get bind
1) "bind"
2) "10.88.0.117"
```
But still the same error
```
10.88.0.117:6666> cluster slots
(error) CLUSTERDOWN The cluster is not initialized
```

**Reply by git-hulk** [reply]
Need to re-init the cluster topo after restarting.

**Reply by rkozlo** [reply]
What you mean by "re-init"? Reset? This command isn't supported so i guess it is not. I even purged whole persistent kvrocks storage.

**Reply by git-hulk** [reply]
My bad, what I mean is to resend `CLUSTERX SETNODES` to nodes.

**Reply by rkozlo** [reply]
But still the same error as on beginning. BTW i have redis in another container and after starting container i can user cluster slots

**Reply by git-hulk** [reply]
Other users also suffered this issue, we can use redis-cli to allocate slots, you can have a try at： https://github.com/apache/incubator-kvrocks/blob/unstable/tools/try_cluster/try_cluster.sh

**Reply by rkozlo** [reply]
@git-hulk tried with 2.2.0 and now error is `(error) Slot is out of range`

**Reply by git-hulk** [reply]
The valid slot range is 0-16383. Can you paste your clusterx setnodes command? It should be something with this command argument.

**Reply by rkozlo** [reply]
```
127.0.0.1:6666> clusterx setnodeid kvrockskvrockskvrockskvrockskvrocksnode1
OK
127.0.0.1:6666> clusterx setnodes "kvrockskvrockskvrockskvrockskvrocksnode1 10.32.68.251 6666 master - 0-5460  kvrockskvrockskvrockskvrockskvrocksnode2 10.32.68.250 6667 master - 5461-10992  kvrockskvrockskvrockskvrockskvrocksnode3 10.32.68.249 6666 master - 10993-16383" 1
(error) Slot is out of range
```
Tried config from try_cluster
```
127.0.0.1:6666> clusterx setnodes "kvrockskvrockskvrockskvrockskvrocksnode1 127.0.0.1 30001 master - 0-5460 kvrockskvrockskvrockskvrockskvrocksnode2 127.0.0.1 30002 slave kvrockskvrockskvrockskvrockskvrocksnode1 kvrockskvrockskvrockskvrockskvrocksnode3 127.0.0.1 30003 master - 5461-10992  kvrockskvrockskvrockskvrockskvrocksnode4 127.0.0.1 30004 slave kvrockskvrockskvrockskvrockskvrocksnode3  kvrockskvrockskvrockskvrockskvrocksnode5 127.0.0.1 30005 master - 10993-16383 kvrockskvrockskvrockskvrockskvrocksnode6 127.0.0.1 30006 slave kvrockskvrockskvrockskvrockskvrocksnode5" 1
(error) Slot is out of range
```
weird thing is in some random moments it passes
```
127.0.0.1:6666> clusterx setnodes "kvrockskvrockskvrockskvrockskvrocksnode1 10.32.68.251 6666 master - 0-5460" 1
OK
127.0.0.1:6666> clusterx setnodes "kvrockskvrockskvrockskvrockskvrocksnode1 10.32.68.251 6666 master - 0-5460  kvrockskvrockskvrockskvrockskvrocksnode2 10.32.68.250 6667 master - 5461-10992  kvrockskvrockskvrockskvrockskvrocksnode3 10.32.68.249 6666 master - 10993-16383" 1
OK
127.0.0.1:6666> cluster slots
1) 1) (integer) 0
   2) (integer) 5460
   3) 1) "10.32.68.251"
      2) (integer) 6666
      3) "kvrockskvrockskvrockskvrockskvrocksnode1"
127.0.0.1:6666>
```
am i doing something wrong?

**Reply by git-hulk** [reply]
Thanks @rkozlo 

For the first question, should add a new line between nodes:

```shell
clusterx setnodes "kvrockskvrockskvrockskvrockskvrocksnode1 10.32.68.251 6666 master - 0-5460 \n kvrockskvrockskvrockskvrockskvrocksnode2 10.32.68.250 6667 master - 5461-10992 \n kvrockskvrockskvrockskvrockskvrocksnode3 10.32.68.249 6666 master - 10993-16383" 1
```

And for the second question, it works well since the version is the same as the previous command and won't check the node's format, so didn't return the error.

**Reply by rkozlo** [reply]
Ok, it works. One question and i'm out ;)
```
127.0.0.1:6666> cluster help
(error) ERR CLUSTER command, CLUSTER INFO|NODES|SLOTS|KEYSLOT
127.0.0.1:6666> clusterx help
(error) ERR CLUSTERX command, CLUSTERX VERSION|SETNODEID|SETNODES|SETSLOT|MIGRATE
```
only those options for `cluster` and `clusterx`  are supported? No failover?

**Reply by git-hulk** [reply]
Yes, the failover for the Kvrocks cluster is to change the cluster topology. It's expected to do outside the kvrocks like [Kvrocks Controller](https://github.com/KvrocksLabs/kvrocks_controller)(working in progress now)

**Reply by rkozlo** [reply]
@git-hulk  thanks for help, you can close this topic, because i can't

**Reply by git-hulk** [reply]
@rkozlo Thanks for your feedback too. Sure, let's lock this discussion.
